src/prove.rs

```diff
@@ -1,88 +1,223 @@
+
+use std::{collections::HashMap, ops::Range, path::Path};
+use streaming_iterator::StreamingIterator;
+
+use crate::{config::Config, difficulty::proving_difficulty, metadata, reader::stream_data};
 
 const BLOCK_SIZE: usize = 16; // size of the aes block
-const BATCH: usize = 8; // will use encrypt8 asm method
-const CHUNK_SIZE: usize = BLOCK_SIZE * BATCH;
+const AES_BATCH: usize = 8; // will use encrypt8 asm method
+const CHUNK_SIZE: usize = BLOCK_SIZE * AES_BATCH;
 
-pub struct Prover<const N: usize = 1> {
-    ciphers: [Aes128; N],
-    output: [u8; CHUNK_SIZE],
-    d: u64,
+#[derive(Debug)]
+pub struct Proof {
+    pub nonce: u32,
+    pub indicies: Vec<u64>,
 }
 
-impl<const N: usize> Prover<N> {
-    pub fn new(challenge: &[u8; 16], d: u64) -> Self {
-        let ciphers: [Aes128; N] = (0..N as u32)
-            .map(|i: u32| {
-                let mut key = [0u8; BLOCK_SIZE];
-                key[..12].copy_from_slice(&challenge[..12]);
-                key[12..].copy_from_slice(&i.to_le_bytes());
-                Aes128::new(&key.into())
-            })
-            .collect::<Vec<_>>()
-            .try_into()
-            .unwrap();
-        let output = [0u8; CHUNK_SIZE];
-        Prover { ciphers, output, d }
+pub trait Prover<'a> {
+    fn prove<F>(&self, batch: &[u8], index: u64, consume: F) -> Option<u32>
+    where
+        F: FnMut(u32, u64) -> bool;
+
+    fn required_aeses(&self) -> usize;
+}
+
+#[derive(Debug)]
+struct AesCipher {
+    pub aes: Aes128,
+    pub nonce: u32,
+}
+
+impl AesCipher {
+    fn new(challenge: &[u8; 32], nonce: u32) -> Self {
+        let key_cipher = Aes256::new(challenge.into());
+        let mut key = [0u8; BLOCK_SIZE];
+        key[0..4].copy_from_slice(&nonce.to_le_bytes());
+        key_cipher.encrypt_block(&mut key.into());
+        Self {
+            aes: Aes128::new(&key.into()),
+            nonce,
+        }
     }
+}
+
+pub struct ConstDProver {
+    ciphers: Vec<AesCipher>,
+    difficulty: u64,
+}
 
-    pub fn prove<F>(&mut self, stream: &[u8], mut consume: F)
+impl ConstDProver {
+    pub fn new(challenge: &[u8; 32], difficulty: u64, nonces: Range<u32>) -> Self {
+        let start = nonces.start / 2;
+        let end = 1.max(nonces.end / 2);
+        ConstDProver {
+            ciphers: (start..end).map(|n| AesCipher::new(challenge, n)).collect(),
+            difficulty,
+        }
+    }
+}
+
+impl<'a> Prover<'a> for ConstDProver {
+    fn required_aeses(&self) -> usize {
+        self.ciphers.len()
+    }
+
+    fn prove<F>(&self, batch: &[u8], mut index: u64, mut consume: F) -> Option<u32>
     where
-        F: FnMut(u64, u64) -> Result<(), ()>,
+        F: FnMut(u32, u64) -> bool,
     {
-        for i in 0..stream.len() / (CHUNK_SIZE) {
-            let chunk = &stream[i * CHUNK_SIZE..(i + 1) * CHUNK_SIZE];
-            for (j, cipher) in self.ciphers.iter().enumerate() {
+        let mut u64s = [0u64; CHUNK_SIZE / 8];
+
+        for chunk in batch.chunks(CHUNK_SIZE) {
```

Comment:
regular loop was faster compared to chunks

Reply:
I don't think it should be - I get the same results. `std::clice::chunks` has the benefit that
- it cannot panic when the index is out of range, unlike the indexed access.
- it's easier to write correctly and read later.

Anyway - I looked up [std::slice::chunks_exact](https://doc.rust-lang.org/std/primitive.slice.html#method.chunks_exact) which is supposed to be faster and it speeds proving up by around ~3%.